File: models/standard_set.rb
require 'pp'
require 'dry-validation'
require 'date'
require_relative '../lib/cache_standards'
require_relative '../lib/send_to_algolia'
require_relative "standard"
require_relative "../lib/securerandom"

class StandardSet
  include Virtus.model

  attribute :id, String, default: -> (page, attrs) { SecureRandom.csp_uuid() }
  attribute :title, String, default: ""
  attribute :subject, String, default: ""
  attribute :document, Hash, default: {}
  attribute :createdAt, DateTime, default: -> (page, attrs) { Time.now}
  attribute :updatedAt, DateTime
  attribute :version, Integer, default: 1
  attribute :standards, Hash[String => Standard], default: {}
  attribute :standardsCount, Integer, default: 0
  attribute :normalizedSubject, String, default: nil

  class Jurisdiction
    include Virtus.model
    attribute :id, String
    attribute :title, String

    class Validator < ::Dry::Validation::Schema
      key(:id, &:str?)
      key(:title, &:str?)
    end
  end
  attribute :jurisdiction, Jurisdiction, default: -> (page, attrs){Jurisdiction.new}

  class CspStatus
    include Virtus.model
    attribute :value, String, default: "visible"
    attribute :notes, String, default: nil
    class Validator < ::Dry::Validation::Schema
      key(:id, &:str?)
      key(:title, &:str?)
    end
  end
  attribute :cspStatus, CspStatus, default: -> (page, attrs){CspStatus.new}


  class License
    include Virtus.model
    attribute :title, String, default: "CC BY 4.0 US"
    attribute :URL, String, default: "http://creativecommons.org/licenses/by/4.0/us/"
    attribute :rightsHolder, String, default: "Common Curriculum, Inc."
    class Validator < ::Dry::Validation::Schema
      key(:title, &:str?)
      key(:URL, &:str?)
      key(:rightsHolder, &:str?)
    end
  end
  attribute :license, License, default: -> (page, attrs){License.new}

  attribute :educationLevels, Array[String], default: []
  EDUCATION_LEVELS = [
    "Pre-K",
    "K",
    "01",
    "02",
    "03",
    "04",
    "05",
    "06",
    "07",
    "08",
    "09",
    "10",
    "11",
    "12",
    "VocationalTraining",
    "ProfessionalEducation-Development",
    "Graduate",
    "HigherEducation",
    "Undergraduate-UpperDivision",
    "Undergraduate-LowerDivision",
    "AdultEducation",
    "LifeLongLearning",
  ]

  class Validator < ::Dry::Validation::Schema
    key(:title, &:str?)
    key(:subject, &:str?)
    key(:educationLevels) {|attr| attr.empty? | attr.in_education_levels?}
    def in_education_levels?(value)
      # the intersection of the array should produce an array of the same length
      # as given if all the given education levels are in the EDUCATION_LEVELS set
      (EDUCATION_LEVELS & value).length == value.length
    end
  end

  def self.validate(model)
    # pp model.attributes
    self_validations = Validator.new.call(model.attributes)
    return self_validations.messages unless self_validations.messages.empty?

    jurisdiction_validations = Jurisdiction::Validator.new.call(model.jurisdiction.attributes)
    return jurisdiction_validations.messages unless jurisdiction_validations.messages.empty?

    license_validations = License::Validator.new.call(model.license.attributes)
    return license_validations.messages unless license_validations.messages.empty?

    return true
  end

  def self.find(id)
    attrs = $db[:standard_sets].find({_id: id}).first
    self.from_mongo(attrs)
  end

  def self.update(doc, opts={})
    old_version    = $db[:standard_sets].find({_id: doc[:id]}).to_a.first || {}
    if old_version
      self.save_version(old_version)
    end

    # Set the version
    doc[:version] = old_version[:version] || 0
    doc[:version] = doc[:version] + 1
    doc[:updatedAt] = DateTime.now()

    id = doc[:id]

    # Set the ID
    doc.delete(:id)
    standards = doc[:standards]
    doc[:standardsCount] = standards&.length || 0

    if id == nil
      pp doc
      raise "_id is nil"
    end

    # Replace the document
    doc = $db[:standard_sets].find({_id: id}).find_one_and_update(doc, {upsert: true, return_document: :after})

    # Cache standards
    unless opts[:cache_standards] == false
      CachedStandards.one(doc)
    end

    # Send to algolia
    unless opts[:send_to_algolia] == false
      SendToAlgolia.standard_set(doc)
    end
  end

  def self.save_version(old_version)
    old_version["standardSetId"] = old_version["_id"]
    old_version["_id"] = SecureRandom.csp_uuid()
    old_version["createdAt"] = Time.now
    $db[:standard_set_versions].insert_one(old_version)
  end

  def self.from_mongo(attrs)
    attrs[:id] = attrs.delete("_id")
    self.new(attrs)
  end


end

File: elixir/lib/csp_api/schemas/standard_set.ex
defmodule CspApi.Schemas.StandardSet do
  @moduledoc """
  Port of `models/standard_set.rb`. Sub-documents use `embeds_one` with
  explicit `@primary_key {:id, :string, autogenerate: false}` so the
  Mongo adapter stores `id` as `id` (it only remaps unkeyed `:id` fields
  to `_id`). The `standards` collection is `:map` because the Ruby model
  is `Hash[id => Standard]` and that's how it's stored on disk.
  """

  use Ecto.Schema
  import Ecto.Changeset

  alias __MODULE__.{Jurisdiction, CspStatus, License, Document}

  @primary_key {:id, :string, autogenerate: false}

  @education_levels ~w(
    Pre-K K 01 02 03 04 05 06 07 08 09 10 11 12
    VocationalTraining ProfessionalEducation-Development Graduate
    HigherEducation Undergraduate-UpperDivision Undergraduate-LowerDivision
    AdultEducation LifeLongLearning
  )

  def education_levels, do: @education_levels

  schema "standard_sets" do
    field :title, :string, default: ""
    field :subject, :string, default: ""
    field :normalizedSubject, :string
    field :educationLevels, {:array, :string}, default: []

    # Standards are stored as a Hash[id => Standard] (matching the Ruby
    # model). The hierarchy walker in `CspApi.Hierarchy` populates
    # `ancestorIds`/`parentId` on read.
    field :standards, :map, default: %{}
    field :standardsCount, :integer, default: 0
    field :version, :integer, default: 1
    field :createdAt, :utc_datetime
    field :updatedAt, :utc_datetime

    embeds_one :document, Document, on_replace: :update
    embeds_one :jurisdiction, Jurisdiction, on_replace: :update
    embeds_one :cspStatus, CspStatus, on_replace: :update
    embeds_one :license, License, on_replace: :update
  end

  @cast_fields ~w(
    id title subject normalizedSubject educationLevels standards
    standardsCount version createdAt updatedAt
  )a

  def changeset(set, attrs) do
    set
    |> cast(attrs, @cast_fields)
    |> cast_embed(:document)
    |> cast_embed(:jurisdiction)
    |> cast_embed(:cspStatus)
    |> cast_embed(:license)
    |> validate_required([:title, :subject])
    |> validate_education_levels()
  end

  defp validate_education_levels(changeset) do
    case get_field(changeset, :educationLevels) do
      nil -> changeset
      [] -> changeset
      levels ->
        if Enum.all?(levels, &(&1 in @education_levels)) do
          changeset
        else
          add_error(changeset, :educationLevels, "contains unknown level")
        end
    end
  end

  defmodule Jurisdiction do
    use Ecto.Schema
    import Ecto.Changeset

    # Explicit string `:id` primary key — without this, mongodb_ecto
    # would treat the `id` field as the embedded doc's auto-renamed
    # primary key and store it as `_id`, breaking `jurisdiction.id`
    # filters and the wire format the Ruby app reads.
    @primary_key {:id, :string, autogenerate: false}
    @derive {Jason.Encoder, only: [:id, :title]}
    embedded_schema do
      field :title, :string
    end

    def changeset(j, attrs) do
      j |> cast(attrs, [:id, :title]) |> validate_required([:id, :title])
    end
  end

  defmodule Document do
    use Ecto.Schema
    import Ecto.Changeset

    @primary_key {:id, :string, autogenerate: false}
    @derive {Jason.Encoder,
             only: [:id, :title, :asnIdentifier, :publicationStatus, :sourceURL, :valid]}
    embedded_schema do
      field :title, :string
      field :asnIdentifier, :string
      field :publicationStatus, :string
      field :sourceURL, :string
      field :valid, :string
    end

    def changeset(d, attrs),
      do: cast(d, attrs, [:id, :title, :asnIdentifier, :publicationStatus, :sourceURL, :valid])
  end

  defmodule CspStatus do
    use Ecto.Schema
    import Ecto.Changeset

    @primary_key false
    @derive {Jason.Encoder, only: [:value, :notes]}
    embedded_schema do
      field :value, :string, default: "visible"
      field :notes, :string
    end

    def changeset(c, attrs), do: cast(c, attrs, [:value, :notes])
  end

  defmodule License do
    use Ecto.Schema
    import Ecto.Changeset

    @primary_key false
    @derive {Jason.Encoder, only: [:title, :URL, :rightsHolder]}
    embedded_schema do
      field :title, :string, default: "CC BY 4.0 US"
      field :URL, :string, default: "http://creativecommons.org/licenses/by/4.0/us/"
      field :rightsHolder, :string, default: "Common Curriculum, Inc."
    end

    def changeset(l, attrs), do: cast(l, attrs, [:title, :URL, :rightsHolder])
  end
end
